x/subspaces/keeper/invariants.go

```diff
@@ -282,3 +286,91 @@ func formatOutputUserPermissions(entries []types.UserPermission) (output string)
 	}
 	return output
 }
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ValidUserGrantsInvariant checks that all the user grants are valid
+func ValidUserGrantsInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var invalidEntries []types.Grant
+		k.IterateUserGrants(ctx, func(entry types.Grant) (stop bool) {
+			invalid := false
+
+			// Check subspace existence
+			if !k.HasSubspace(ctx, entry.SubspaceID) {
+				invalid = true
+			}
+
+			// Validate the entry
+			err := entry.Validate()
+			if err != nil {
+				invalid = true
+			}
+
+			if invalid {
+				invalidEntries = append(invalidEntries, entry)
+			}
+
+			return false
+		})
+
+		return sdk.FormatInvariant(types.ModuleName, "invalid user grants",
+			fmt.Sprintf("the following user grants are invalid:\n%s", formatOutputUserGrants(invalidEntries)),
+		), invalidEntries != nil
+	}
+}
+
+// formatOutputUserGrants concatenates the given user grants into a string
+func formatOutputUserGrants(entries []types.Grant) (output string) {
+	for _, entry := range entries {
+		output += fmt.Sprintf("SubspaceID: %d, Granter: %s, Grantee: %s\n", entry.SubspaceID, entry.Granter, entry.Grantee.GetCachedValue().(*types.UserGrantee).User)
+	}
+	return output
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// ValidGroupGrantsInvariant checks that all the user grants are valid
+func ValidGroupGrantsInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		var invalidEntries []types.Grant
+		k.IterateUserGroupsGrants(ctx, func(entry types.Grant) (stop bool) {
+			invalid := false
+
+			// Check subspace existence
+			if !k.HasSubspace(ctx, entry.SubspaceID) {
+				invalid = true
+			}
+
+			groupID := entry.Grantee.GetCachedValue().(*types.GroupGrantee).GroupID
+			// Check group existence
+			if !k.HasUserGroup(ctx, entry.SubspaceID, groupID) {
```

Comment:
Even though more verbose, I think it makes it easier to read to split the retrieval of the grantee and the usage of its fields:
```suggestion
			// Check group existence
			grantee := entry.Grantee.GetCachedValue().(*types.GroupGrantee)
			if !k.HasUserGroup(ctx, entry.SubspaceID, grantee.GroupID) {
```